manuals/luksdemount.1
.Dd July  1, 2026
.Dt LUKSDEMOUNT 1
.Os
.Sh NAME
.Nm luksdemount
.Nd mounts a Linux Unified Key Setup (LUKS) Disk Encrypted volume
.Sh SYNOPSIS
.Nm luksdemount
.Op Fl k Ar key
.Op Fl o Ar offset
.Op Fl p Ar password
.Op Fl X Ar extended_options
.Op Fl huvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm luksdemount
is a utility to mount a Linux Unified Key Setup (LUKS) Disk Encrypted volume
.Pp
.Nm luksdemount
is part of the
.Nm libluksde
package.
.Nm libluksde
is a library to access the Linux Unified Key Setup (LUKS) Disk Encryption format
.Pp
.Ar source
a Linux Unified Key Setup (LUKS) Disk Encrypted volume
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl k Ar key
specify the key formatted in base16
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl p Ar password
specify the password (or passphrase)
.It Fl u
unattended mode (disables user interaction)
.It Fl v
verbose output to stderr, while luksdemount will remain running in the \
foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# luksdemount image1.raw /mnt/luksde
luksdemount 20260701
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr luksdeinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libluksde/issues
.Sh COPYRIGHT
Copyright (C) 2013-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* add more password hashing support
  - fix caluclation of validation hash for SHA-224 and SHA-256 password hash
* documentation
  - add SHA224 hashing method
* add thread-safety in volume
  - add tests
* detect encryption types (notify as unsupported)
* add libluksde_diffuser_split?

* library: add support to set the master key directly

Crypto support:
* add support for lmk
* add support for serpent-CBC
* add support for twofish
* add support for cast5
* add support for cast6

Hashing support:
* add support for RIPEMD160
* add support for whirlpool

20141001
* see `git log' for more recent change log
* removed README.macosx
* changes for project site move

20140801
* bug fix in Python-bindings

20140531
* updated dependencies

20140330
* worked on Python bindings

20140112
* 2014 update
* updated dependencies

20131205
* updated dependencies

20130909
* updated dependencies
* worked on libcthreads build support

20130716
* updated dependencies

20130714
* worked on SHA512 support

20130704
* worked on initial version
* renamed libluksve to libluksde
* removed unused local libraries

20130703
* initial version based on libbde



luksdetools/luksdemount.c
/*
 * Mounts a Linux Unified Key Setup (LUKS) Disk Encrypted volume.
 *
 * Copyright (C) 2013-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "luksdetools_getopt.h"
#include "luksdetools_i18n.h"
#include "luksdetools_libcerror.h"
#include "luksdetools_libclocale.h"
#include "luksdetools_libcnotify.h"
#include "luksdetools_libluksde.h"
#include "luksdetools_output.h"
#include "luksdetools_signal.h"
#include "luksdetools_unused.h"
#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"

mount_handle_t *luksdemount_mount_handle = NULL;
int luksdemount_abort                    = 0;

/* Signal handler for luksdemount
 */
void luksdemount_signal_handler(
      luksdetools_signal_t signal LUKSDETOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "luksdemount_signal_handler";

	LUKSDETOOLS_UNREFERENCED_PARAMETER( signal )

	luksdemount_abort = 1;

	if( luksdemount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     luksdemount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use luksdemount to mount a Linux Unified Key Setup (LUKS) Disk Encrypted volume.";

	luksdetools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'k', "key", "specify the key formatted in base16" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'p', "password", "specify the password (or passphrase)" },
		{ 'u', NULL, "unattended mode (disables user interaction)" },
		{ 'v', NULL, "verbose output to stderr, while luksdemount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "volume", "a Linux Unified Key Setup (LUKS) Disk Encrypted volume" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	const system_character_t *path_prefix       = NULL;
	libluksde_error_t *error                    = NULL;
	size_t path_prefix_size                     = 0;
	system_character_t *option_key              = NULL;
	system_character_t *option_offset           = NULL;
	system_character_t *option_password         = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "luksdemount";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( luksdetools_option_t ) );
	int unattended_mode                         = 0;
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
	int result                                  = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations luksdemount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                        = { program, NULL };
	struct fuse_args luksdemount_fuse_arguments = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args luksdemount_fuse_arguments = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *luksdemount_fuse_channel  = NULL;
#endif
	struct fuse *luksdemount_fuse_handle        = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS luksdemount_dokan_operations;
	DOKAN_OPTIONS luksdemount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "luksdetools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( luksdetools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	luksdetools_output_version_fprint(
	 stdout,
	 program );

	if( luksdetools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = luksdetools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				luksdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				luksdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'k':
				option_key = optarg;

				break;

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'u':
				unattended_mode = 1;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				luksdetools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		luksdetools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		luksdetools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libluksde_notify_set_stream(
	 stderr,
	 NULL );
	libluksde_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &luksdemount_mount_handle,
	     unattended_mode,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_key != NULL )
	{
		if( mount_handle_set_key(
		     luksdemount_mount_handle,
		     option_key,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set key.\n" );

			goto on_error;
		}
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     luksdemount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
	if( option_password != NULL )
	{
		if( mount_handle_set_password(
		     luksdemount_mount_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\LUKSDE" );
#else
	path_prefix = _SYSTEM_STRING( "/luksde" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     luksdemount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	if( mount_handle_open(
	     luksdemount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume\n" );

		goto on_error;
	}
	if( mount_handle_is_locked(
	     luksdemount_mount_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to unlock source volume\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		luksdemount_fuse_arguments.argc = 0;
		luksdemount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &luksdemount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &luksdemount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &luksdemount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &luksdemount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	luksdemount_fuse_operations.open       = &mount_fuse_open;
	luksdemount_fuse_operations.read       = &mount_fuse_read;
	luksdemount_fuse_operations.release    = &mount_fuse_release;
	luksdemount_fuse_operations.opendir    = &mount_fuse_opendir;
	luksdemount_fuse_operations.readdir    = &mount_fuse_readdir;
	luksdemount_fuse_operations.releasedir = &mount_fuse_releasedir;
	luksdemount_fuse_operations.getattr    = &mount_fuse_getattr;
	luksdemount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	luksdemount_fuse_handle = fuse_new(
	                           &luksdemount_fuse_arguments,
	                           &luksdemount_fuse_operations,
	                           sizeof( struct fuse_operations ),
	                           luksdemount_mount_handle );

	if( luksdemount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          luksdemount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	luksdemount_fuse_channel = fuse_mount(
	                            mount_point,
	                            &luksdemount_fuse_arguments );

	if( luksdemount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	luksdemount_fuse_handle = fuse_new(
	                           luksdemount_fuse_channel,
	                           &luksdemount_fuse_arguments,
	                           &luksdemount_fuse_operations,
	                           sizeof( struct fuse_operations ),
	                           luksdemount_mount_handle );

	if( luksdemount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          luksdemount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 luksdemount_fuse_handle );

	fuse_destroy(
	 luksdemount_fuse_handle );

	fuse_opt_free_args(
	 &luksdemount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &luksdemount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &luksdemount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	luksdemount_dokan_options.Version    = DOKAN_VERSION;
	luksdemount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	luksdemount_dokan_options.SingleThread = TRUE;
#else
	luksdemount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		luksdemount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		luksdemount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	luksdemount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	luksdemount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	luksdemount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	luksdemount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	luksdemount_dokan_operations.CreateDirectory      = NULL;
	luksdemount_dokan_operations.Cleanup              = NULL;
	luksdemount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	luksdemount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	luksdemount_dokan_operations.WriteFile            = NULL;
	luksdemount_dokan_operations.FlushFileBuffers     = NULL;
	luksdemount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	luksdemount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	luksdemount_dokan_operations.FindFilesWithPattern = NULL;
	luksdemount_dokan_operations.SetFileAttributes    = NULL;
	luksdemount_dokan_operations.SetFileTime          = NULL;
	luksdemount_dokan_operations.DeleteFile           = NULL;
	luksdemount_dokan_operations.DeleteDirectory      = NULL;
	luksdemount_dokan_operations.MoveFile             = NULL;
	luksdemount_dokan_operations.SetEndOfFile         = NULL;
	luksdemount_dokan_operations.SetAllocationSize    = NULL;
	luksdemount_dokan_operations.LockFile             = NULL;
	luksdemount_dokan_operations.UnlockFile           = NULL;
	luksdemount_dokan_operations.GetFileSecurity      = NULL;
	luksdemount_dokan_operations.SetFileSecurity      = NULL;
	luksdemount_dokan_operations.GetDiskFreeSpace     = NULL;
	luksdemount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	luksdemount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	luksdemount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	luksdemount_dokan_operations.Cleanup              = NULL;
	luksdemount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	luksdemount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	luksdemount_dokan_operations.WriteFile            = NULL;
	luksdemount_dokan_operations.FlushFileBuffers     = NULL;
	luksdemount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	luksdemount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	luksdemount_dokan_operations.FindFilesWithPattern = NULL;
	luksdemount_dokan_operations.SetFileAttributes    = NULL;
	luksdemount_dokan_operations.SetFileTime          = NULL;
	luksdemount_dokan_operations.DeleteFile           = NULL;
	luksdemount_dokan_operations.DeleteDirectory      = NULL;
	luksdemount_dokan_operations.MoveFile             = NULL;
	luksdemount_dokan_operations.SetEndOfFile         = NULL;
	luksdemount_dokan_operations.SetAllocationSize    = NULL;
	luksdemount_dokan_operations.LockFile             = NULL;
	luksdemount_dokan_operations.UnlockFile           = NULL;
	luksdemount_dokan_operations.GetFileSecurity      = NULL;
	luksdemount_dokan_operations.SetFileSecurity      = NULL;
	luksdemount_dokan_operations.GetDiskFreeSpace     = NULL;
	luksdemount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	luksdemount_dokan_operations.Unmounted            = NULL;
	luksdemount_dokan_operations.FindStreams          = NULL;
	luksdemount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &luksdemount_dokan_options,
	          &luksdemount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Linux Unified Key Setup (LUKS) Disk Encryption format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( luksdemount_fuse_handle != NULL )
	{
		fuse_destroy(
		 luksdemount_fuse_handle );
	}
	fuse_opt_free_args(
	 &luksdemount_fuse_arguments );
#endif
	if( luksdemount_mount_handle != NULL )
	{
		mount_handle_free(
		 &luksdemount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

